File: python2-sdk/semantria/authrequest.py
# -*- coding: utf-8 -*-
import httplib
import base64
import urllib
import time
import random
import urlparse
import hmac
import binascii
import hashlib

OAuthVersion = "1.0"
OAuthParameterPrefix = "oauth_"
OAuthConsumerKeyKey = "oauth_consumer_key"
OAuthVersionKey = "oauth_version"
OAuthSignatureMethodKey = "oauth_signature_method"
OAuthSignatureKey = "oauth_signature"
OAuthTimestampKey = "oauth_timestamp"
OAuthNonceKey = "oauth_nonce"

class AuthRequest:
    def __init__(self, consumerKey, consumerSecret, applicationName):
        self.consumerKey = consumerKey
        self.consumerSecret = consumerSecret
        self.applicationName = applicationName;

    def authWebRequest(self, method, url, postData):
        nonce = self.generateNonce()
        timestamp = self.generateTimestamp()
        query = self.generateQuery(method, url, timestamp, nonce)
        authheader = self.generateAuthHeader(query, timestamp, nonce)

        headers = {"Authorization": authheader}
        if (method == "POST"):
            headers["Content-type"] = "application/x-www-form-urlencoded"
        headers["x-api-version"] = "2"
        headers["x-app-name"] = self.applicationName
        
        qparts = urlparse.urlparse(query)
        qscheme, qnetloc, qpath, qparams, qquery, qfragment = qparts[:6]
        conn = httplib.HTTPSConnection(qnetloc)
        host = '%s?%s' % (qpath, qquery)
        conn.request(method, host, postData, headers)
        response = conn.getresponse()
        result = { "status":response.status, "reason":response.reason, "data":response.read() }
        conn.close()
        return result

    def generateQuery(self, method, url, timestamp, nonce):
        parts = urlparse.urlparse(url)
        scheme, netloc, path, params, query, fragment = parts[:6]
        np = self.getNormalizedParameters(timestamp, nonce)

        if query:
            query = '%s&oauth_verifier=%s' % (query, np)
        else:
            query = '%s' % np
        
        return urlparse.urlunparse((scheme, netloc, path, params, query, fragment))         

    def generateAuthHeader(self, query, timestamp, nonce):
        md5cs = self.getMD5Hash(self.consumerSecret)
        escquery = self.escape(query)
        hash = self.getSHA1(md5cs, escquery)
        hash = self.escape(hash)

        items = []
        items.append(("OAuth realm", ""))
        items.append((OAuthVersionKey, "%s" % OAuthVersion))
        items.append((OAuthTimestampKey, "%s" % timestamp))
        items.append((OAuthNonceKey, "%s" % nonce))
        items.append((OAuthSignatureMethodKey, "HMAC-SHA1"))
        items.append((OAuthConsumerKeyKey, "%s" % self.consumerKey))
        items.append((OAuthSignatureKey, "%s" % hash))
        items.sort()

        params = []
        for k, v in items:
            params.append('%s="%s"' % (k, v))

        return ",".join(params)

    def getNormalizedParameters(self, timestamp, nonce):
        items = []
        items.append((OAuthVersionKey, OAuthVersion))
        items.append((OAuthTimestampKey, timestamp))
        items.append((OAuthNonceKey, nonce))
        items.append((OAuthSignatureMethodKey, "HMAC-SHA1"))
        items.append((OAuthConsumerKeyKey, self.consumerKey))
        items.sort()
        
        encoded_str = urllib.urlencode(items)
        # Encode signature parameters per Oauth Core 1.0 protocol
        # Spaces must be encoded with "%20" instead of "+"
        return encoded_str.replace('+', '%20').replace('%7E', '~')

    def getMD5Hash(self, str):
        #bytestr = bytearray(str) 
        md5h = hashlib.md5()
        #md5h.update(bytestr)
        md5h.update(str)
        md5sig = md5h.hexdigest()
        return md5sig
 
    def getSHA1(self, md5cs, query):
        sha1res = hmac.new(md5cs, query, hashlib.sha1).digest()
        sha1sig = binascii.b2a_base64(sha1res)[:-1]
        return sha1sig

    def generateTimestamp(self):
        return int(time.time())

    def generateNonce(self, length=20):
        return ''.join([str(random.randint(0, 9)) for i in range(length)])
    
    def escape(self, s):
        return urllib.quote(s.encode('utf-8'), safe='~')

File: Go/authrequest.go
package sem4

import (
	"crypto/hmac"
	"crypto/md5"
	"crypto/sha1"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"math/rand"
	"net/url"
	"sort"
	"strings"
	"time"

	"github.com/levigross/grequests"
	"github.com/rs/zerolog/log"
)

const (
	authorizationHeader     = "Authorization"
	oAuthVersion            = "1.0"
	oAuthConsumerKeyKey     = "oauth_consumer_key"
	oAuthVersionKey         = "oauth_version"
	oAuthSignatureMethodKey = "oauth_signature_method"
	oAuthSignatureKey       = "oauth_signature"
	oAuthTimestampKey       = "oauth_timestamp"
	oAuthNonceKey           = "oauth_nonce"
)

// AuthRequest represents an authenticated request to Semantria
type AuthRequest struct {
	ConsumerKey        string
	ConsumerSecret     string
	ApplicationName    string
	APIVersion         string
	DisableCompression bool
}

// AuthRequestOptions represents the options to a particular authenticated request
type AuthRequestOptions struct {
	// JSON can be used for sending data
	JSON interface{}

	// Headers if you want to add additional headers to the request
	Headers map[string]string
}

// DoAuthRequest makes an authenticated request
func (ar AuthRequest) DoAuthRequest(verb string, url string, options *AuthRequestOptions) (string, error) {
	if options == nil {
		options = &AuthRequestOptions{}
	}

	if !DisableLogging {
        log.Debug().Msgf("%s %s", verb, url)
	}

	nonce := generateNonce(20)
	timestamp := ar.generateTimestamp()
	requestURL, err := generateAuthURL(verb, url, ar.ConsumerKey, timestamp, nonce)
	if err != nil {
		return "", err
	}
	authorization := generateAuthHeader(requestURL, ar.ConsumerKey, ar.ConsumerSecret, timestamp, nonce)
	authHeaders := ar.generateHeaders(verb, authorization, options.Headers)

	goptions := &grequests.RequestOptions{
		Headers:            authHeaders,
		DisableCompression: ar.DisableCompression,
	}
	if options.JSON != nil {
		goptions.JSON = options.JSON
	}

	resp, err := grequests.Req(verb, requestURL, goptions)
	if err != nil {
		return "", nil
	}
	if !resp.Ok {
		return "", errors.New("Request failed: " + resp.String())
	}
	if resp.StatusCode == 202 {
		return "", nil
	}
	return resp.String(), nil
}

// AuthGet makes an authenticate GET request
func (ar AuthRequest) AuthGet(url string, options *AuthRequestOptions) (string, error) {
	return ar.DoAuthRequest("GET", url, options)
}

// AuthPost makes an authenticate GET request
func (ar AuthRequest) AuthPost(url string, options *AuthRequestOptions) (string, error) {
	return ar.DoAuthRequest("POST", url, options)
}

// generateNonce generates the random nonce
func generateNonce(length int) string {
	r := rand.New(rand.NewSource(time.Now().Unix()))
	s := ""
	for n := 0; n < length; n++ {
		s += fmt.Sprintf("%d", r.Intn(10))
	}
	return s
}

func (ar AuthRequest) generateTimestamp() int64 {
	return time.Now().Unix()
}

// generateAuthURL adds the oath magic got the url
func generateAuthURL(verb string, requestURL string, consumerKey string, timestamp int64, nonce string) (string, error) {
	parts, err := url.Parse(requestURL)
	if err != nil {
		return "", err
	}
	np := getNormalizedParameters(consumerKey, timestamp, nonce)

	authURL := requestURL
	if len(parts.RawQuery) > 0 {
		authURL += "&oauth_verifier=" + url.QueryEscape(np)
	} else {
		authURL += "?" + np
	}
	return authURL, nil
}

// getSHA1 calculates the hmac digest
func getSHA1(key string, message string) string {
	mac := hmac.New(sha1.New, []byte(key))
	mac.Write([]byte(message))
	sha1res := mac.Sum(nil)
	sha1sig := base64.StdEncoding.EncodeToString(sha1res)
	return sha1sig
}

// generateMD5FromString creates an MD5 digest from the given message
func generateMD5FromString(message string) string {
	h := md5.New()
	io.WriteString(h, message)
	return hex.EncodeToString(h.Sum(nil))
}

// escape escapes the url properly
func escape(theURL string) string {
	return url.QueryEscape(theURL)
}

// generateAuthHeader creates the oath authorization value
func generateAuthHeader(requestURL string, consumerKey string, consumerSecret string, timestamp int64, nonce string) string {
	md5cs := generateMD5FromString(consumerSecret)
	escapedURL := escape(requestURL)
	hash := escape(getSHA1(md5cs, escapedURL))

	pairs := []pair{
		{"OAuth realm", ""},
		{oAuthVersionKey, oAuthVersion},
		{oAuthTimestampKey, fmt.Sprintf("%d", timestamp)},
		{oAuthNonceKey, nonce},
		{oAuthSignatureMethodKey, "HMAC-SHA1"},
		{oAuthConsumerKeyKey, consumerKey},
		{oAuthSignatureKey, hash},
	}
	sort.Sort(byKey(pairs))

	var builder strings.Builder
	for _, pair := range pairs {
		if builder.Len() > 0 {
			builder.WriteString(",")
		}
		builder.WriteString(pair.Key)
		builder.WriteString(`="`)
		builder.WriteString(pair.Value)
		builder.WriteString(`"`)
	}
	return builder.String()
}

// pair is a pair
type pair struct {
	Key   string
	Value string
}

// byKey implements sorting by the key
type byKey []pair

// Len is the length of A
func (a byKey) Len() int {
	return len(a)
}

// Less tells if i is less than j
func (a byKey) Less(i, j int) bool {
	return a[i].Key < a[j].Key
}

// Swap swaps i for j
func (a byKey) Swap(i, j int) {
	a[i], a[j] = a[j], a[i]
}

// getNormalizedParameters creates the oath parameters string
func getNormalizedParameters(consumerKey string, timestamp int64, nonce string) string {
	pairs := []pair{
		{oAuthVersionKey, oAuthVersion},
		{oAuthTimestampKey, fmt.Sprintf("%d", timestamp)},
		{oAuthNonceKey, nonce},
		{oAuthSignatureMethodKey, "HMAC-SHA1"},
		{oAuthConsumerKeyKey, consumerKey},
	}
	sort.Sort(byKey(pairs))

	encodedStr := ""
	for _, pair := range pairs {
		if len(encodedStr) > 0 {
			encodedStr += "&"
		}
		encodedStr += url.QueryEscape(pair.Key)
		encodedStr += "="
		encodedStr += url.QueryEscape(fmt.Sprintf("%s", pair.Value))
	}
	// Encode signature parameters per Oauth Core 1.0 protocol
	// Spaces must be encoded with "%20" instead of "+"
	encodedStr = strings.ReplaceAll(encodedStr, "+", "%20")
	encodedStr = strings.ReplaceAll(encodedStr, "%7E", "~")

	return encodedStr
}

func (ar AuthRequest) generateHeaders(verb string, authorization string, customHeaders map[string]string) map[string]string {
	headers := make(map[string]string)
	if customHeaders != nil {
		for key, value := range customHeaders {
			headers[key] = value
		}
	}
	headers[authorizationHeader] = authorization
	if ar.APIVersion != "" {
		headers["x-api-version"] = ar.APIVersion
	} else {
		headers["x-api-version"] = DefaultAPIVersion
	}
	if ar.ApplicationName != "" {
		headers["x-app-name"] = ar.ApplicationName
	}
	return headers
}
